### DISCUSSION TITLE
Automatically generating and using filenames based on date, animal ID, and experiment type

### DISCUSSION THREAD

**Discussion by KangByungwoo** [bonsai-rx/bonsai#1552] [Q&A]
Hi,

I'm wondering if there is a way to automatically generate and use filenames based on date, animal ID, and experiment type and etc, as manually doing this is cumbersome. I found a similar discussion (https://github.com/orgs/bonsai-rx/discussions/1308) and tried to modify the sample codes at https://bonsai-rx.org/docs/articles/expressions-format.html?q=Format, but I couldn't really find a solution as I'm not very experienced in bonsai. If someone could give detailed suggestions or a sample code I can modify to achieve this, that would be really helpful. Below is my current workflow, and I'd like to change filenames in both CSV writers and video writers.

![image](https://github.com/bonsai-rx/bonsai/assets/26239966/529674c8-cf4d-418f-9e87-ec51cc7f20fe)

**Comment by bruno-f-cruz** [comment]
What happens when you try the solution suggested in the `Format` documentation?

**Reply by KangByungwoo** [reply]
I wasn't able to add more than one input node for Format node. Is there a way to do it?

**Reply by bruno-f-cruz** [reply]
The easist way is to use `Tuples`:

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/d8ddbb1e-cb92-452c-b6bb-50eeb472a28c)

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="StringProperty">
          <Value>String</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="IntProperty">
          <Value>13123</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Zip" />
      </Expression>
      <Expression xsi:type="Format">
        <Format>This is a {0} and the number {1} </Format>
        <Selector>Item1, Item2</Selector>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="2" Label="Source1" />
      <Edge From="1" To="2" Label="Source2" />
      <Edge From="2" To="3" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>

**Comment by sstaszko** [comment]
I have a similar question - the format node works nicely, except that it populates my various write notes with the most recently used animal/date information. So, if i run the script twice, I get the correct updated formatting, but if I run it once, I get the prior formatting. The workflow is attached. I appreciate any suggestions!

[Single LifeCam Optogenetics Keypress_delayedstart.zip](https://github.com/user-attachments/files/16732984/Single.LifeCam.Optogenetics.Keypress_delayedstart.zip)

**Reply by bruno-f-cruz** [reply]
Make sure you dont branch, as this creates racing conditions as described in https://bonsai-rx.org/docs/articles/workflow-guidelines.html#property-initialization


Follow the pattern recommended in the docs
https://bonsai-rx.org/docs/articles/expressions-format.html?q=format
